semantic-core/src/Data/Core/Pretty.hs

```diff
@@ -31,88 +31,91 @@ showFile = showCore . fileBody
+              , symbol "=" <+> align (withPrec 0 (go (instantiate1 (pure (name x)) b)))
+              ]
+
+            Lam (Named (Ignored x) b) -> prec 3 . group . nest 2 $ vsep
+              [ lambda <> name x, arrow <+> withPrec 0 (go (instantiate1 (pure (name x)) b)) ]
+
+            Record fs -> atom . group . nest 2 $ vsep [ primitive "record", block ", " (map (uncurry keyValue) fs) ]
+
+            Unit     -> atom $ primitive "unit"
+            Bool b   -> atom $ primitive (if b then "true" else "false")
+            String s -> atom . strlit $ viaShow s
+
+            f :$ x -> prec 8 (withPrec 8 (go f) <+> withPrec 9 (go x))
+
+            If con tru fal -> prec 3 . group $ vsep
+              [ keyword "if"   <+> precBody (go con)
+              , keyword "then" <+> precBody (go tru)
+              , keyword "else" <+> precBody (go fal)
+              ]
+
+            Load p -> prec 3 (keyword "load" <+> withPrec 9 (go p))
+            item :. body -> prec 9 (withPrec 9 (go item) <> symbol "." <> name body)
+
+            lhs := rhs -> prec 3 . group . nest 2 $ vsep
+              [ withPrec 4 (go lhs)
+              , symbol "=" <+> align (withPrec 4 (go rhs))
+              ]
 
             -- Annotations are not pretty-printed, as it lowers the signal/noise ratio too profoundly.
             Ann _ c -> go c
-          where bind (Ignored x) f = (,) x <$> go (instantiate1 (pure x) f)
+            statement ->
+              let (bindings, return) = unstatements (Term statement)
+                  statements = toList (bindings :> (Nothing :<- return))
+                  names = zipWith (\ i (n :<- _) -> maybe (pretty @Int i) (name . namedName) n) [0..] statements
+                  statements' = map (prettyStatement names) statements
+              in atom (block "; " statements')
+        block _ [] = braces mempty
+        block s ss = encloseSep "{ " " }" s ss
+        keyValue x v = name x <+> symbol ":" <+> precBody (go v)
+        prettyStatement names (Just (Named (Ignored u) _) :<- t) = name u <+> arrowL <+> precBody (go (either (names !!) id <$> t))
+        prettyStatement names (Nothing                    :<- t) = precBody (go (either (names !!) id <$> t))
         lambda = case style of
           Unicode -> symbol "λ"
           Ascii   -> symbol "\\"
         arrow = case style of
           Unicode -> symbol "→"
           Ascii   -> symbol "->"
+        arrowL = case style of
+          Unicode -> symbol "←"
+          Ascii   -> symbol "<-"
+
+
+data Prec a = Prec
+  { precLevel :: Maybe Int
+  , precBody  :: a
+  }
+  deriving (Eq, Ord, Show)
+
+prec :: Int -> a -> Prec a
+prec = Prec . Just
 
+atom :: a -> Prec a
+atom = Prec Nothing
 
-appending :: Functor f => AnsiDoc -> f AnsiDoc -> f AnsiDoc
-appending k item = (keyword k <+>) <$> item
+withPrec :: Int -> Prec AnsiDoc -> AnsiDoc
+withPrec d (Prec d' a)
+  | maybe False (d >) d' = parens a
+  | otherwise            = a
```

Comment:
I rewrote the pretty-printer to use the `Prec` datatype because of a number of bugs where the use of a `Reader` effect meant that we were simply passing precedences along or otherwise getting them wrong. We can’t use the `AnsiDoc` inside a `Prec AnsiDoc`, so unwrapping (typically with `withPrec` here) is enforced, and we thus have to specify a precedence (and thus automatically parenthesize as necessary).

This would have been much more complicated were we not handling sequences of statements as a unit, since we would sometimes need to wrap expressions in braces instead of parens.

Reply:
Loving this. Definitely an improvement on my first stabs at it.